Source/WebCore/accessibility/AccessibilityNodeObject.cpp

```diff
@@ -2521,6 +2629,261 @@ AXCoreObject* AccessibilityNodeObject::parentTableIfExposedTableRow() const
+        if (elementName == ElementName::HTML_tfoot || elementName == ElementName::HTML_tbody || elementName == ElementName::HTML_table) {
+            // If we're in the first column, we're a row header.
+            if (!columnIndexRange().first)
+                return true;
+            return false;
+        }
+
+        if (elementName == ElementName::HTML_thead)
+            return false;
+    }
+    return false;
+}
+
+std::pair<unsigned, unsigned> AccessibilityNodeObject::rowIndexRange() const
+{
+    ensureIndexesUpToDate();
+    return hasRareData() ? std::make_pair(rareData()->rowIndex(), rareData()->effectiveRowSpan()) : std::make_pair(0u, 1u);
+}
+
+std::pair<unsigned, unsigned> AccessibilityNodeObject::columnIndexRange() const
+{
+    ensureIndexesUpToDate();
+    return hasRareData() ? std::make_pair(rareData()->columnIndex(), colSpan()) : std::make_pair(0u, 1u);
+}
+
+String AccessibilityNodeObject::axColumnIndexText() const
+{
+    return getAttribute(aria_colindextextAttr);
+}
+
+unsigned AccessibilityNodeObject::colSpan() const
+{
+    if (!isTableCell())
+        return 1;
+
+    if (auto colSpan = parseHTMLInteger(getAttribute(colspanAttr)); colSpan && *colSpan >= 1) {
+        // https://html.spec.whatwg.org/multipage/tables.html
+        // If colspan is greater than 1000, let it be 1000 instead.
+        return std::min(std::max(*colSpan, 1), 1000);
+    }
+    if (auto ariaColSpan = parseHTMLInteger(getAttribute(aria_colspanAttr)); ariaColSpan && *ariaColSpan >= 1)
+        return std::min(std::max(*ariaColSpan, 1), 1000);
+    return 1;
+}
+
+unsigned AccessibilityNodeObject::rowSpan() const
+{
+    if (!isTableCell())
+        return 1;
+    // According to the ARIA spec, "If aria-rowspan is used on an element for which the host language
+    // provides an equivalent attribute, user agents must ignore the value of aria-rowspan."
+    if (auto rowSpan = parseHTMLInteger(getAttribute(rowspanAttr))) {
+        if (*rowSpan < 0)
+            return 1;
+        return std::min(static_cast<unsigned>(*rowSpan), HTMLTableCellElement::maxRowspan);
+    }
+
+    if (auto ariaRowSpan = parseHTMLInteger(getAttribute(aria_rowspanAttr))) {
+        if (*ariaRowSpan < 0)
+            return 1;
+        return std::min(static_cast<unsigned>(*ariaRowSpan), HTMLTableCellElement::maxRowspan);
+    }
+
+    return 1;
+}
+
+void AccessibilityNodeObject::incrementEffectiveRowSpan()
+{
+    if (hasRareData())
+        rareData()->incrementEffectiveRowSpan();
+}
+
+void AccessibilityNodeObject::resetEffectiveRowSpan()
+{
+    if (hasRareData())
+        rareData()->resetEffectiveRowSpan();
+}
+
+void AccessibilityNodeObject::setAXColIndexFromRow(int index)
+{
+    if (!hasRareData() && index == -1)
+        return;
+    ensureRareData().setAXColIndexFromRow(index);
+}
+
+void AccessibilityNodeObject::setColumnIndex(unsigned index)
+{
+    if (!isTableCell())
+        return;
+
+    if (!hasRareData() && !index)
+        return;
```

Comment:
Could we leave a comment explaining that this is an optimization (I presume I'm understanding that correctly?) to avoid creating a rare data when we don't need to? Or am I reading this wrong? This took me a little while to wrap my head around

Reply:
I also kind of wonder whether something that's a table cell is going to always have rare data created for it anyways through some other field, making this check a bit unnecessary? Or do you think it's reasonable that we'd have a number of cells benefit from this in common table layouts?